src/boot/fwupd.c

```diff
@@ -0,0 +1,589 @@
+
+                _cleanup_(fwup_update_table_freep) FwupUpdateTable *update = NULL;
+                err = parse_info_var(name, &update);
+                if (err != EFI_SUCCESS) {
+                        (void) efivar_unset(MAKE_GUID_PTR(FWUPDATE), name, /* flags= */ 0);
+                        return err;
+                }
+                if (!FLAGS_SET(update->info->status, FWUPDATE_ATTEMPT_UPDATE))
+                        continue;
+
+                if (*n_updates == SIZE_MAX)
+                        return log_error_status(EFI_OUT_OF_RESOURCES, "Too many UEFI update variables found");
+
+                *updates = xrealloc(*updates,
+                                    *n_updates * sizeof(FwupUpdateTable),
+                                    (*n_updates + 1) * sizeof(FwupUpdateTable));
+                (*updates)[(*n_updates)++] = TAKE_PTR(update);
+        }
+
+        return EFI_SUCCESS;
+}
+
+static EFI_STATUS get_graphics_protocol_mode(EFI_HANDLE image, uint32_t *ret_mode) {
+        _cleanup_free_ EFI_HANDLE *handles = NULL;
+        size_t n_handles = 0;
+        static uint32_t mode = 0;
+        static int found = -1;
+        EFI_STATUS err;
+
+        assert(ret_mode);
+
+        /* Doesn't depend on the capsule being inspected, cache the result */
+        if (found > 0) {
+                *ret_mode = mode;
+                return EFI_SUCCESS;
+        }
+        if (found == 0)
+                return EFI_UNSUPPORTED;
+
+        err = BS->LocateHandleBuffer(
+                        ByProtocol,
+                        MAKE_GUID_PTR(EFI_GRAPHICS_OUTPUT_PROTOCOL),
+                        /* SearchKey= */ NULL,
+                        &n_handles,
+                        &handles);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Failed to locate graphics output protocol handles: %m");
+        if (n_handles == 0 || !handles)
+                return log_error_status(EFI_UNSUPPORTED, "No graphics output protocol handles found for UX capsule");
+
+        FOREACH_ARRAY(h, handles, n_handles) {
+                EFI_GRAPHICS_OUTPUT_PROTOCOL *proto;
+
+                err = BS->OpenProtocol(*h,
+                                       MAKE_GUID_PTR(EFI_GRAPHICS_OUTPUT_PROTOCOL),
+                                       (void **)&proto,
+                                       image,
+                                       /* ControllerHandler= */ NULL,
+                                       EFI_OPEN_PROTOCOL_GET_PROTOCOL);
+                if (err != EFI_SUCCESS)
+                        return log_error_status(err, "Failed to get graphics output protocol: %m");
+
+                *ret_mode = mode = proto->mode->mode;
+                found = 1;
+                return EFI_SUCCESS;
+        }
+
+        found = 0;
+        return EFI_UNSUPPORTED;
+}
+
+static EFI_STATUS device_path_find_index_of_type_and_truncate(EFI_DEVICE_PATH *dp, size_t *ret_index) {
```

Comment:
i don't grok what this func name has to do with what the function does? it seems to look for MEDIA_DEVICE_PATH and truncate there